[4.4] Fix ordering for files in the template view

HLeithner opened the pull request:
Pull Request for Issue #40605 .

### Summary of Changes
Order is now using the right values for displaying directory structure for templates in the backend.
This could happen on posix filesystems because the natural order is the file creation and not the alphabetical order.

### Testing Instructions
Check if your directories and files are sorted in the corrected order


### Actual result BEFORE applying this Pull Request

The order was only for directories correct


### Expected result AFTER applying this Pull Request

Sorting happens in the model and not longer in the view for directories and files.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### Additional
![image](https://github.com/joomla/joomla-cms/assets/1497730/e8911496-e283-4159-83c5-fae55fc2c087)

HLeithner committed:
Fix ordering for files in the template view

dgrammatiko commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -430,6 +430,24 @@ public function getDirectoryTree($dir)
+            }
+
+            if (\is_array($b)) {
+
```
```suggestion
```

dautrich commented:
@HLeithner 
The fix is for J4, isn't it? Can I apply the patch to J5.1 as well?

brianteeman commented:
Is the image a before or after. If after then it is showing the folders in the tree not being in alphabetical order

HLeithner commented:
I think so, I didn't see much changes since 4.4 in 5.1

dautrich commented:
I have tested this item :white_check_mark: successfully on f3a8732a925047651965d23533928c1afb5c4a13<br /><br />I successfully tested the PR on a J5.1.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43335">issues.joomla.org/tracker/joomla-cms/43335</a>.</sub>

HLeithner committed:
Rewrite getDirectoryTree function using early return and uksort

HLeithner commented:
> Is the image a before or after. If after then it is showing the folders in the tree not being in alphabetical order

correct php fooled me and the over complicated php code didn't make it better. I updated the code and moved from uasort to uksort. Now it works as expected.

Here the image from joomla 5.1:
![image](https://github.com/joomla/joomla-cms/assets/1497730/fade5d7a-5923-4907-bff0-95308f1013c3)

HLeithner commented:
> Now I am really confused. The original PR used uasort and did not work with cassiopeia. Yet @dautrich says it does work with the j51 template

it partly works with cassiopeia as you pointed out based on the image (media folder -> images shouldn't be on the last place, btw. this information in you comment would have saved me 20 minutes...). so his test is wrong.

HLeithner committed:
CS

dautrich commented:
### Template Skylar

![grafik](https://github.com/joomla/joomla-cms/assets/7933215/cc0a826f-d166-41c5-ad3d-01ecea405d9c)

brianteeman commented:
Please review https://github.com/joomla/joomla-cms/pull/25792/

HLeithner committed:
Use natural sorting instead of string comparison

HLeithner commented:
> Please review #25792

done

dautrich commented:
@HLeithner 
Do you want me to re-test this?

HLeithner commented:
> @HLeithner Do you want me to re-test this?

Plz

dautrich commented:
I have tested this item :white_check_mark: successfully on 4edaa4c1d0b730e9ef021f55c99eeedef48ea6f2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43335">issues.joomla.org/tracker/joomla-cms/43335</a>.</sub>

dautrich commented:
![43335 - 2024-04-29](https://github.com/joomla/joomla-cms/assets/7933215/8dab593f-0d9d-464f-b257-f9ea8f26a47b)

viocassel commented:
I have tested this item :white_check_mark: successfully on 4edaa4c1d0b730e9ef021f55c99eeedef48ea6f2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43335">issues.joomla.org/tracker/joomla-cms/43335</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43335">issues.joomla.org/tracker/joomla-cms/43335</a>.</sub>

QuyTon committed:
Merge branch '4.4-dev' into 4.4/fix/template-file-ordering

richard67 committed:
Merge branch '4.4-dev' into 4.4/fix/template-file-ordering

MacJoom committed:
Merge branch '4.4-dev' into 4.4/fix/template-file-ordering

MacJoom commented:
Thank you!
